### PR TITLE
Add CBMC proofs for handleUnexpectedEvents

### PR DESCRIPTION
**Pull request by Bujain**
Add CBMC proofs for handleUnexpectedEvents function

Description
-----------
This PR adds files required for the CBMC proof of handleUnexpectedEvents function. 

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] My code is formatted using Uncrustify.
- [x] I have read and applied the rules stated in CONTRIBUTING.md. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/include/stubs.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                         const void * pData );
+
+/* Stub to reset the OTA PAL layer. */
+OtaPalStatus_t resetPalStub( OtaFileContext_t * const pFileContext );
+
+/* Stub to set the Image State. */
+OtaErr_t setImageStateWithReason( OtaImageState_t stateToSet,
+                                  uint32_t reasonToSet );
+
+OtaErr_t requestJobStub( OtaAgentContext_t * pAgentCtx );
```
Comment (gshvang) [inline test/cbmc/include/stubs.h:54]: Comment describing what these functions are doing will be helpful.

File: test/cbmc/source/stubs.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    /* state must have values of OtaPalImageState_t enum. */
+    __CPROVER_assume( ( state >= OtaPalImageStateUnknown ) && ( state <= OtaPalImageStateInvalid ) );
+
+    __CPROVER_assert( pFileContext != NULL,
+                      "Error: pFileContext in the otaAgent is statically initialized and hence cannot be NULL." );
+
+    return state;
+}
+
+void otaAppCallback( OtaJobEvent_t eEvent,
```
Comment (gshvang) [inline test/cbmc/source/stubs.c:83]: Empty return can be added instead of leaving the body empty.

File: test/cbmc/source/stubs.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    __CPROVER_assert( pTimerName != NULL, "Error: TimerName cannot be NULL" );
+
+    __CPROVER_assert( callback != NULL, "Error: callback should be pointing to a function." );
+
+    return status;
+}
+
+OtaOsStatus_t stopTimerStub( OtaTimerId_t otaTimerId )
+{
+    OtaOsStatus_t status;
```
Comment (gshvang) [inline test/cbmc/source/stubs.c:62]: Initiailization of status should be done.

File: test/cbmc/source/stubs.c [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    return err;
+}
+
+OtaOsStatus_t startTimerStub( OtaTimerId_t otaTimerId,
+                              const char * const pTimerName,
+                              const uint32_t timeout,
+                              OtaTimerCallback_t callback )
+{
+    OtaOsStatus_t status;
```
Comment (gshvang) [inline test/cbmc/source/stubs.c:48]: Initiailization of status should be done.

File: test/cbmc/source/stubs.c (hunk 2 above)
Comment (Bujain) [reply test/cbmc/source/stubs.c:83]: Uncrustify does not accept empty return and changes it to a empty body.

**Review by gshvang — APPROVED** [review]

**Review by gshvang — DISMISSED** [review]

File: test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+void handleUnexpectedEvents_harness()
+{
+    OtaEventMsg_t * pEventMsg;
+    OtaEvent_t eventId;
+
+    pEventMsg = ( OtaEventMsg_t * ) malloc( sizeof( OtaEventMsg_t ) );
+
+    /* pEventMsg cannot be NULL as it is statically initialized in receiveAndProcessOtaEvent
+     * before handleUnexpectedEvents function is called. */
+    __CPROVER_assume( pEventMsg != NULL );
```
Comment (paulbartell) [inline test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c:44]: Perhaps there should be an equivalent assertion in handleUnexpectedEvents to cover this assumption.

File: test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    /* eventId can only assume values of OtaEvent_t enum. */
+    __CPROVER_assume( ( eventId >= OtaAgentEventStart ) && ( eventId <= OtaAgentEventMax ) );
```
Comment (paulbartell) [inline test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c:47]: This assumption is not really necessary given the current implementation.

File: test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+extern OtaAgentContext_t otaAgent;
+extern void handleUnexpectedEvents( const OtaEventMsg_t * pEventMsg );
+
+void handleUnexpectedEvents_harness()
+{
+    OtaEventMsg_t * pEventMsg;
+    OtaEvent_t eventId;
+
+    pEventMsg = ( OtaEventMsg_t * ) malloc( sizeof( OtaEventMsg_t ) );
```
Comment (paulbartell) [inline test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c:40]: Is there a particular reason that this is heap allocated?

File: test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    /* eventId can only assume values of OtaEvent_t enum. */
+    __CPROVER_assume( ( eventId >= OtaAgentEventStart ) && ( eventId <= OtaAgentEventMax ) );
+
+    pEventMsg->eventId = eventId;
+
+    /* Initialize OtaAppCallback to an empty callback function. */
+    otaAgent.OtaAppCallback = otaAppCallbackStub;
+
+    ( void ) handleUnexpectedEvents( pEventMsg );
```
Comment (paulbartell) [inline test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c:54]: The results of the call to handleUnexpectedEvents with eventID = OtaAgentEventReceivedFileBlock is not validated.
https://github.com/aws/ota-for-aws-iot-embedded-sdk/blob/82bac55e867a971f68706f0c42d8780e5aa4b6bf/source/ota.c#L2796

File: test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c (hunk 7 above)
Comment (Bujain) [reply test/cbmc/proofs/handleUnexpectedEvents/handleUnexpectedEvents_harness.c:40]: No, there is no reason for it to be heap allocated. I've changed it to static allocation.

**Review by AniruddhaKanhere — DISMISSED** [review]

**Review by paulbartell — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Pull request merged**